egs/wsj/s5/steps/libs/nnet3/xconfig/parser.py

```diff
@@ -67,11 +67,73 @@ def xconfig_line_to_object(config_line, prev_layers = None):
             "*** {0}".format(config_line))
         raise
 
+# This function reads existing model (*.raw or *.mdl) and returns array of
+# XconfigInputLayer one per input-node or component-node with same 'name' used
+# in raw model and 'dim' equal to 'output-dim' for component-node and 'dim' for
+# input-node.
+def get_model_component_info(model_filename):
+    all_layers = []
+    try:
+        f = open(model_filename, 'r')
+    except Exception as e:
+        sys.exit("{0}: error reading model file '{1}'".format(sys.argv[0], model_filename, repr(e)))
+
+    # use nnet3-info to get component names in the model.
+    out = common_lib.get_command_stdout("""nnet3-info {0} | grep '\-node' """
+                                        """ """.format(model_filename))
+
+    # out contains all {input,component}-nodes used in model_filename
+    # It can parse lines in out like:
+    # i.e. input-node name=input dim=40
+    #      component-node name=tdnn1.affine component=tdnn1.affine input=lda input-dim=300 output-dim=512
+    layer_names = []
+    for line in out.split("\n"):
+        parts = line.split(" ")
+        input_dim = -1
+        output_dim = -1
+        for  field in parts:
+            key_value = field.split("=")
+            if len(key_value) == 2:
+                key = key_value[0]
+                value = key_value[1]
+                if key == "name":           # name=**
+                    layer_name = value
+                elif key == "dim":          # for input-node
+                    input_dim = int(value)
```

Comment:
you should be setting output_dim here, because logically an input node has no input, only an output.
that would enable some of the code below to be simplified.  In fact, I don't see why you need separate input_dim and output_dim variables; you could just have a single 'dim' which is the output dim of the layer, and simplify the code quite a bit.